v2/gcs-to-splunk/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSplunkTest.java

```diff
@@ -0,0 +1,311 @@
+
+    GCSToSplunkOptions options = PipelineOptionsFactory.create().as(GCSToSplunkOptions.class);
+
+    options.setJsonSchemaPath(JSON_SCHEMA_FILE_PATH);
+    options.setContainsHeaders(false);
+    options.setInputFileSpec(NO_HEADER_CSV_FILE_PATH);
+
+    // Act
+    PCollectionTuple readCsvOut = readFromCsvAndTransform(options);
+
+    // Assert
+    PAssert.that(readCsvOut.get(UDF_OUT))
+        .satisfies(
+            collection -> {
+              FailsafeElement element = collection.iterator().next();
+              assertThat(element.getPayload(), is(equalTo(stringifiedJsonRecord)));
+              return null;
+            });
+
+    // Act
+    PCollectionTuple convertToSplunkEventOut = convertToSplunkEvent(readCsvOut);
+
+    // Assert
+    PAssert.that(convertToSplunkEventOut.get(SPLUNK_EVENT_OUT))
+        .containsInAnyOrder(expectedSplunkEvent);
+
+    //  Execute pipeline
+    pipeline.run();
+  }
+
+  /** Tests the {@link GCSToSplunk} pipeline when there are write errors. */
+  @Test
+  public void testGCSToSplunkWriteErrors() {
+    // Arrange
+    final String stringifiedError =
+        "Payload: test-payload. Error Message: test-message. Splunk write status code: 123.";
+    CoderRegistry coderRegistry = pipeline.getCoderRegistry();
+    coderRegistry.registerCoderForType(
+        FAILSAFE_ELEMENT_CODER.getEncodedTypeDescriptor(), FAILSAFE_ELEMENT_CODER);
+
+    String payload = "test-payload";
+    String message = "test-message";
+    Integer statusCode = 123;
+
+    SplunkWriteError splunkWriteError =
+        SplunkWriteError.newBuilder()
+            .withPayload(payload)
+            .withStatusCode(statusCode)
+            .withStatusMessage(message)
+            .create();
+
+    // Act
+    PCollection<SplunkWriteError> splunkErrorCollection =
+        pipeline.apply(
+            "Create Input data", Create.of(splunkWriteError).withCoder(SplunkWriteErrorCoder.of()));
+
+    // Assert
+    PAssert.that(splunkErrorCollection).containsInAnyOrder(splunkWriteError);
+
+    // Act
+    PCollection<FailsafeElement<String, String>> wrappedSplunkWriteErrors =
+        wrapSplunkErrorsToFailsafe(splunkErrorCollection);
+
+    // Assert
+    PAssert.that(wrappedSplunkWriteErrors)
+        .satisfies(
+            collection -> {
+              FailsafeElement element = collection.iterator().next();
+              assertThat(element.getPayload(), is(equalTo(payload)));
+              assertThat(
+                  element.getErrorMessage(),
+                  is(equalTo(message + ". Splunk write status code: " + statusCode)));
+              return null;
+            });
+
+    // Act
+    PCollectionTuple flattenedErrors = getStringifiedFailsafeErrors(wrappedSplunkWriteErrors);
+
+    // Assert
+    PAssert.that(flattenedErrors.get(COMBINED_ERRORS)).containsInAnyOrder(stringifiedError);
+
+    //  Execute pipeline
+    pipeline.run();
+  }
+
+  private PCollectionTuple readFromCsvAndTransform(GCSToSplunkOptions options) {
```

Comment:
It looks like this method was copied over from the production code. The pipeline code should be structured in a way to be tested directly. Requiring a copy/paste is error-prone and confusing for future developers. Same goes for the other cases.